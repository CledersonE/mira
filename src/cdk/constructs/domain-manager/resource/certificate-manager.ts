import { AccountPrincipal, Effect, ManagedPolicy, PolicyStatement, Role, ServicePrincipal } from '@aws-cdk/aws-iam'
import { Code, Runtime, SingletonFunction } from '@aws-cdk/aws-lambda'
import { SnsEventSource } from '@aws-cdk/aws-lambda-event-sources'
import { Topic } from '@aws-cdk/aws-sns'
import { CfnOutput, Construct, Duration } from '@aws-cdk/core'
import path from 'path'
import { MiraConfig } from '../../../../config/mira-config'
import { MiraStack } from '../../../stack'

export const CERTIFICATE_SUBSCRIPTION_TOPIC = 'CertificateSubscriptionTopic'

/**
 * Certificate manager
 */
export class CertificateManager extends MiraStack {
  constructor (parent: Construct) {
    super(parent, 'CertificateManager')
    const account = MiraConfig.getEnvironment()
    const domainAccount = MiraConfig.getEnvironment('domain')
    const { hostedZoneId } = MiraConfig.getDomainConfig()

    if (!hostedZoneId) {
      throw new Error('"hostedZoneId" is required in "domain" config')
    }

    // Create role for CertificateManager and add policies
    const certificateManagerRole = new Role(this, 'CertificateManagerLambdaRole', {
      assumedBy: new ServicePrincipal('lambda.amazonaws.com')
    })

    certificateManagerRole.addManagedPolicy(ManagedPolicy.fromAwsManagedPolicyName('service-role/AWSLambdaBasicExecutionRole'))

    certificateManagerRole.addToPolicy(new PolicyStatement({
      effect: Effect.ALLOW,
      resources: ['*'],
      actions: ['acm:RequestCertificate', 'acm:DescribeCertificate', 'acm:DeleteCertificate', 'acm:ListCertificates']
    }))

    certificateManagerRole.addToPolicy(new PolicyStatement({
      effect: Effect.ALLOW,
      resources: [`arn:aws:iam::${domainAccount.env.account}:role/${MiraConfig.getBaseStackName('CrossAccountDomainManagerRole')}`],
      actions: ['sts:AssumeRole']
    }))

    // Create SNS topic
    const certificateSubscriptionTopic = new Topic(this, CERTIFICATE_SUBSCRIPTION_TOPIC, {
      displayName: 'Certificate Subscription Topic',
      topicName: MiraConfig.calculateSharedResourceName(CERTIFICATE_SUBSCRIPTION_TOPIC)
    })

    // Allow account to publish message to topic
    certificateSubscriptionTopic.addToResourcePolicy(new PolicyStatement({
      principals: [new AccountPrincipal(account.env.account)],
      effect: Effect.ALLOW,
      resources: [certificateSubscriptionTopic.topicArn],
      actions: ['sns:Publish']
    }))

    // Lambda handler - TODO update code
    const certificateManagerLambda = new SingletonFunction(this, 'CertificateManagerLambda', {
      code: Code.fromAsset(path.join(__dirname, '../lambda')),
      handler: 'certificate-manager.handler',
      runtime: Runtime.NODEJS_12_X,
      timeout: Duration.minutes(15),
      uuid: '59d1148b-c159-4a58-883e-fd31df31a2d9',
      environment: {
        HOSTED_ZONE: hostedZoneId
      },
      role: certificateManagerRole
    })

    certificateManagerLambda.addEventSource(new SnsEventSource(certificateSubscriptionTopic))

    new CfnOutput(this, 'certificateSubscriptionTopicArn', {
      value: certificateSubscriptionTopic.topicArn
    })
  }
}
