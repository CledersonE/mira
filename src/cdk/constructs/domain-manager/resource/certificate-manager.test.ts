import { SynthUtils } from "@aws-cdk/assert"
import { Stack, App } from "@aws-cdk/core"

import {
  MiraConfig,
  DomainConfig,
  Account,
} from "../../../../config/mira-config"

import { CertificateManager } from "./certificate-manager"

describe("CertificateManager", () => {
  it("throws when hostedZoneId is not set in config", () => {
    const stack = new Stack(
      new App(),
      MiraConfig.getBaseStackName("CertificateManager"),
      {}
    )

    MiraConfig.getDomainConfig = (): DomainConfig => ({
      accounts: [],
    })

    MiraConfig.getEnvironment = (): Account => ({
      name: "some-name",
      profile: "some-profile",
      env: { account: "12345", region: "eu-west-1" },
    })

    expect(() => new CertificateManager(stack)).toThrowError(
      '"hostedZoneId" is required in "domain" config'
    )
  })

  test("renders output correctly", () => {
    const stack = new Stack(
      new App(),
      MiraConfig.getBaseStackName("CertificateManager"),
      {}
    )
    MiraConfig.getDomainConfig = (): DomainConfig => ({
      hostedZoneId: 'Z123456789',
      accounts: [
        'account1',
        'domain'
      ],
    })

    MiraConfig.getEnvironment = (envName: string): Account => {
      if (envName === 'domain') {
        return ({
          name: "domain",
          profile: "domain",
          env: { account: "domain", region: "eu-west-1" },
        })
      }

      return {
        name: "account1",
        profile: "account1",
        env: { account: "account1", region: "eu-west-1" },
      }
    }
    const cm = new CertificateManager(stack)
    expect(SynthUtils.toCloudFormation(cm)).toMatchSnapshot()
  })
})
