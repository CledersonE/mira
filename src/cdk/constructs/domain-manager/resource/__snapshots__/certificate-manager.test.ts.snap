// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CertificateManager renders output correctly 1`] = `
Object {
  "Outputs": Object {
    "certificateSubscriptionTopicArn": Object {
      "Value": Object {
        "Ref": "CertificateSubscriptionTopic416ACB99",
      },
    },
  },
  "Parameters": Object {
    "referencetoJohnMyGreatAppCertificateManagerAssetParameters67a2713d64d20eb121a60fd742aa52e95e70ba6b3e2ca98eae7a2b004382af9dS3Bucket540E7416Ref": Object {
      "Type": "String",
    },
    "referencetoJohnMyGreatAppCertificateManagerAssetParameters67a2713d64d20eb121a60fd742aa52e95e70ba6b3e2ca98eae7a2b004382af9dS3VersionKeyDE3C7046Ref": Object {
      "Type": "String",
    },
  },
  "Resources": Object {
    "CertificateManagerLambdaCertificateSubscriptionTopic4E6C3270": Object {
      "Properties": Object {
        "Endpoint": Object {
          "Fn::GetAtt": Array [
            "SingletonLambda59d1148bc1594a58883efd31df31a2d97B5E3F8F",
            "Arn",
          ],
        },
        "Protocol": "lambda",
        "TopicArn": Object {
          "Ref": "CertificateSubscriptionTopic416ACB99",
        },
      },
      "Type": "AWS::SNS::Subscription",
    },
    "CertificateManagerLambdaRole363A1081": Object {
      "Properties": Object {
        "AssumeRolePolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": Object {
                "Service": "lambda.amazonaws.com",
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "ManagedPolicyArns": Array [
          Object {
            "Fn::Join": Array [
              "",
              Array [
                "arn:",
                Object {
                  "Ref": "AWS::Partition",
                },
                ":iam::aws:policy/service-role/AWSLambdaBasicExecutionRole",
              ],
            ],
          },
        ],
      },
      "Type": "AWS::IAM::Role",
    },
    "CertificateManagerLambdaRoleDefaultPolicy2FD77DAD": Object {
      "Properties": Object {
        "PolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": Array [
                "acm:RequestCertificate",
                "acm:DescribeCertificate",
                "acm:DeleteCertificate",
                "acm:ListCertificates",
              ],
              "Effect": "Allow",
              "Resource": "*",
            },
            Object {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Resource": "arn:aws:iam::domain:role/John-MyGreatApp-CrossAccountDomainManagerRole",
            },
          ],
          "Version": "2012-10-17",
        },
        "PolicyName": "CertificateManagerLambdaRoleDefaultPolicy2FD77DAD",
        "Roles": Array [
          Object {
            "Ref": "CertificateManagerLambdaRole363A1081",
          },
        ],
      },
      "Type": "AWS::IAM::Policy",
    },
    "CertificateSubscriptionTopic416ACB99": Object {
      "Properties": Object {
        "DisplayName": "Certificate Subscription Topic",
        "TopicName": "John-MyGreatApp-CertificateSubscriptionTopic",
      },
      "Type": "AWS::SNS::Topic",
    },
    "CertificateSubscriptionTopicPolicyC3EA145B": Object {
      "Properties": Object {
        "PolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": "sns:Publish",
              "Effect": "Allow",
              "Principal": Object {
                "AWS": Object {
                  "Fn::Join": Array [
                    "",
                    Array [
                      "arn:",
                      Object {
                        "Ref": "AWS::Partition",
                      },
                      ":iam::account1:root",
                    ],
                  ],
                },
              },
              "Resource": Object {
                "Ref": "CertificateSubscriptionTopic416ACB99",
              },
              "Sid": "0",
            },
          ],
          "Version": "2012-10-17",
        },
        "Topics": Array [
          Object {
            "Ref": "CertificateSubscriptionTopic416ACB99",
          },
        ],
      },
      "Type": "AWS::SNS::TopicPolicy",
    },
    "SingletonLambda59d1148bc1594a58883efd31df31a2d97B5E3F8F": Object {
      "DependsOn": Array [
        "CertificateManagerLambdaRoleDefaultPolicy2FD77DAD",
        "CertificateManagerLambdaRole363A1081",
      ],
      "Properties": Object {
        "Code": Object {
          "S3Bucket": Object {
            "Ref": "referencetoJohnMyGreatAppCertificateManagerAssetParameters67a2713d64d20eb121a60fd742aa52e95e70ba6b3e2ca98eae7a2b004382af9dS3Bucket540E7416Ref",
          },
          "S3Key": Object {
            "Fn::Join": Array [
              "",
              Array [
                Object {
                  "Fn::Select": Array [
                    0,
                    Object {
                      "Fn::Split": Array [
                        "||",
                        Object {
                          "Ref": "referencetoJohnMyGreatAppCertificateManagerAssetParameters67a2713d64d20eb121a60fd742aa52e95e70ba6b3e2ca98eae7a2b004382af9dS3VersionKeyDE3C7046Ref",
                        },
                      ],
                    },
                  ],
                },
                Object {
                  "Fn::Select": Array [
                    1,
                    Object {
                      "Fn::Split": Array [
                        "||",
                        Object {
                          "Ref": "referencetoJohnMyGreatAppCertificateManagerAssetParameters67a2713d64d20eb121a60fd742aa52e95e70ba6b3e2ca98eae7a2b004382af9dS3VersionKeyDE3C7046Ref",
                        },
                      ],
                    },
                  ],
                },
              ],
            ],
          },
        },
        "Environment": Object {
          "Variables": Object {
            "HOSTED_ZONE": "Z123456789",
          },
        },
        "Handler": "certificate-manager.handler",
        "Role": Object {
          "Fn::GetAtt": Array [
            "CertificateManagerLambdaRole363A1081",
            "Arn",
          ],
        },
        "Runtime": "nodejs12.x",
        "Timeout": 900,
      },
      "Type": "AWS::Lambda::Function",
    },
    "SingletonLambda59d1148bc1594a58883efd31df31a2d9AllowInvokeJohnMyGreatAppCertificateManagerCertificateManager1CertificateSubscriptionTopicA7B1F978F021ED56": Object {
      "Properties": Object {
        "Action": "lambda:InvokeFunction",
        "FunctionName": Object {
          "Fn::GetAtt": Array [
            "SingletonLambda59d1148bc1594a58883efd31df31a2d97B5E3F8F",
            "Arn",
          ],
        },
        "Principal": "sns.amazonaws.com",
        "SourceArn": Object {
          "Ref": "CertificateSubscriptionTopic416ACB99",
        },
      },
      "Type": "AWS::Lambda::Permission",
    },
  },
}
`;
